drivers/dma/dma_gd32.c

```diff
@@ -0,0 +1,626 @@
+		gd32_dma_transfer_set_memory_to_memory(cfg->reg, channel);
+		memory_cfg = &dst_cfg;
+		periph_cfg = &src_cfg;
+		break;
+	case PERIPHERAL_TO_MEMORY:
+		gd32_dma_transfer_set_periph_to_memory(cfg->reg, channel);
+		memory_cfg = &dst_cfg;
+		periph_cfg = &src_cfg;
+		break;
+	case MEMORY_TO_PERIPHERAL:
+		gd32_dma_transfer_set_memory_to_periph(cfg->reg, channel);
+		memory_cfg = &src_cfg;
+		periph_cfg = &dst_cfg;
+		break;
+	}
+
+	gd32_dma_memory_address_config(cfg->reg, channel, memory_cfg->addr);
+	if (memory_cfg->adj == DMA_ADDR_ADJ_INCREMENT) {
+		gd32_dma_memory_increase_enable(cfg->reg, channel);
+	} else {
+		gd32_dma_memory_increase_disable(cfg->reg, channel);
+	}
+
+	gd32_dma_periph_address_config(cfg->reg, channel, periph_cfg->addr);
+	if (periph_cfg->adj == DMA_ADDR_ADJ_INCREMENT) {
+		gd32_dma_periph_increase_enable(cfg->reg, channel);
+	} else {
+		gd32_dma_periph_increase_disable(cfg->reg, channel);
+	}
+
+	gd32_dma_transfer_number_config(cfg->reg, channel, dma_cfg->head_block->block_size);
+	gd32_dma_priority_config(cfg->reg, channel, dma_gd32_priority(dma_cfg->channel_priority));
+	gd32_dma_memory_width_config(cfg->reg, channel, dma_gd32_memory_width(memory_cfg->adj));
+	gd32_dma_periph_width_config(cfg->reg, channel, dma_gd32_periph_width(periph_cfg->adj));
+	gd32_dma_circulation_disable(cfg->reg, channel);
+#ifdef CONFIG_SOC_SERIES_GD32F4XX
+	gd32_dma_channel_subperipheral_select(cfg->reg, channel, dma_cfg->linked_channel);
+#endif
+
+	data->channels[channel].callback = dma_cfg->dma_callback;
+	data->channels[channel].user_data = dma_cfg->user_data;
+	data->channels[channel].direction = dma_cfg->channel_direction;
+
+	return 0;
+}
+
+static int dma_gd32_reload(const struct device *dev, uint32_t ch, uint32_t src, uint32_t dst,
+			   size_t size)
+{
+	const struct dma_gd32_config *cfg = dev->config;
+	struct dma_gd32_data *data = dev->data;
+
+	if (ch >= cfg->channels) {
+		LOG_ERR("reload channel must be < %" PRIu32 " (%" PRIu32 ")", cfg->channels, ch);
+		return -EINVAL;
+	}
+
+	if (data->channels[ch].busy) {
+		return -EBUSY;
+	}
+
+	gd32_dma_channel_disable(cfg->reg, ch);
+
+	gd32_dma_transfer_number_config(cfg->reg, ch, size);
+
+	switch (data->channels[ch].direction) {
+	case MEMORY_TO_MEMORY:
+	case PERIPHERAL_TO_MEMORY:
+		gd32_dma_memory_address_config(cfg->reg, ch, dst);
+		gd32_dma_periph_address_config(cfg->reg, ch, src);
+		break;
+	case MEMORY_TO_PERIPHERAL:
+		gd32_dma_memory_address_config(cfg->reg, ch, src);
+		gd32_dma_periph_address_config(cfg->reg, ch, dst);
+		break;
+	}
+
+	gd32_dma_channel_enable(cfg->reg, ch);
+
+	return 0;
+}
+
+static int dma_gd32_start(const struct device *dev, uint32_t ch)
+{
+	const struct dma_gd32_config *cfg = dev->config;
+	struct dma_gd32_data *data = dev->data;
+
+	if (ch >= cfg->channels) {
+		LOG_ERR("start channel must be < %" PRIu32 " (%" PRIu32 ")", cfg->channels, ch);
+		return -EINVAL;
+	}
+
+	gd32_dma_interrupt_enable(cfg->reg, ch, DMA_FLAG_FTF | GD32_DMA_FLAG_ERRORS);
```

Comment:
For gd32f4xx series, interrupt enable and flag not located in the same bit.
Please consider to use DMA_INTF_xxxIF and DMA_CHXCTL_xxxE seperately.

Reply:
split defines for dma_interrupt_(enable|disable) and dma_interrupt_flag_(set|get)